### ISSUE TITLE
Make `basePath` configurable

### ISSUE THREAD

**Issue by jan-molak** [TypeStrong/typedoc#1924]
## Search terms

Monorepo, lerna, nrwl, fileName

## Expected Behavior

Given a mono-repo with modules under a common directory, for example:

```
packages
- module-A
  - src
    - index.ts
- module-B
  - src
    - index.ts
```

I'd expect `fileName` discovered for `index.ts` in `module-A` (and `module-B`) to include the common directory name, so `packages/module-A/src/index.ts` instead of just `module-A/src/index.ts`

## Actual Behavior

To my understanding, when TypeDoc [parses the sourcecode](https://github.com/TypeStrong/typedoc/blob/e37db58e8e610c77406b0ec083a2b5d8ea42ea0c/src/lib/converter/plugins/SourcePlugin.ts#L127-L135), it calls [`getCommonDirectory`](https://github.com/TypeStrong/typedoc/blob/e37db58e8e610c77406b0ec083a2b5d8ea42ea0c/src/lib/utils/fs.ts#L8) to automatically determine the root directory.

However, for mono-repos that follow the structure recommended by Lerna (with all the public modules under `packages`), the `fileName` determined by TypeDoc is incorrect since it misses the top-level sub-directory name.

This then has a knock-on effect on other projects such as `docusaurus-plugin-typedoc-api`, which use `fileName` to determine the location of the file in the GitHub repository - see https://github.com/milesj/docusaurus-plugin-typedoc-api/issues/39

## Steps to reproduce the bug

See above

## Proposed solution

It _feels like_ making the [`basePath` in `onBeginResolve`](https://github.com/TypeStrong/typedoc/blob/e37db58e8e610c77406b0ec083a2b5d8ea42ea0c/src/lib/converter/plugins/SourcePlugin.ts#L127-L135) configurable could solve this problem:

```diff
    private onBeginResolve(context: Context) {
-        this.basePath = getCommonDirectory([...this.fileNames]);
+        this.basePath = this.basePath || getCommonDirectory([...this.fileNames]);
        for (const file of context.project.files) {
            const fileName = (file.fileName = normalizePath(
                relative(this.basePath, file.fileName)
            ));
            this.fileMappings[fileName] = file;
        }
    }
```

## Environment

-   Typedoc version: `0.22.13`
-   TypeScript version: `4.6.3`
-   Node.js version: `16.13.1`
-   OS: macOS 12.3.1

**Comment by Gerrit0** [comment]
The specific problem you're running into I believe will be resolved by 0.23, which adds a `url` to sources:

https://github.com/TypeStrong/typedoc/blob/03d948ee9c68e47885e8629901ef0e5441a3ce1a/src/test/converter/alias/specs.json#L22-L29

That said, I'm not opposed to a baseDir option which controls this, preventing the getCommonDirectory behavior.

**Comment by milesj** [comment]
@Gerrit0 Will this `url` be available for the JSON outputs also?

**Comment by Gerrit0** [comment]
Yes, url is included in the json in 0.23

**Comment by milesj** [comment]
Awesome, looking forward to it.
